tests/models/test_utils.py

```diff
@@ -321,36 +321,80 @@ def test_enforce_property_with_errors():
         )
 
 
-def test_enforce_array():
-    data = ["some_sentence1", "some_sentence2"]
-    arr = Array(DataType.string)
-    assert _enforce_array(data, arr) == data
-
-    data = np.array(["some_sentence1", "some_sentence2"])
-    assert _enforce_array(data, Array(DataType.string)) == ["some_sentence1", "some_sentence2"]
-
-    data = [
-        {"a": "some_sentence1", "b": "some_sentence2"},
-        {"a": "some_sentence3", "c": ["some_sentence4", "some_sentence5"]},
-    ]
-    arr = Array(
-        Object(
+@pytest.mark.parametrize(
+    ("data", "schema"),
+    [
+        # 1. Simply array
+        (["some_sentence1", "some_sentence2"], Array(DataType.string)),
+        # 2. Numpy
+        (np.array(["some_sentence1", "some_sentence2"]), Array(DataType.string)),
+        # 3. Array of Object
+        (
             [
-                Property("a", DataType.string),
-                Property("b", DataType.string, required=False),
-                Property("c", Array(DataType.string), required=False),
-            ]
-        )
-    )
-    assert _enforce_array(data, arr) == data
+                {"a": "some_sentence1", "b": "some_sentence2"},
+                {"a": "some_sentence3", "c": ["some_sentence4", "some_sentence5"]},
+            ],
+            Array(
+                Object(
+                    [
+                        Property("a", DataType.string),
+                        Property("b", DataType.string, required=False),
+                        Property("c", Array(DataType.string), required=False),
+                    ]
+                )
+            ),
+        ),
+        # 4. Nested array
+        (
+            [
+                [["a", "b"], ["c", "d"]],
+                [["e", "f", "g"], ["h"]],
+                [[]],
+            ],
+            Array(Array(Array(DataType.string))),
+        ),
+        # 5. Numpy 2D array
+        (
+            np.array(
+                [
+                    ["a", "b"],
+                    ["c", "d"],
+                ]
+            ),
+            Array(Array(DataType.string)),
+        ),
+        # 6. Empty array
+        (
+            [],
+            Array(DataType.string),
+        ),
+    ],
+)
+def test_enforce_array(data, schema):
```

Comment:
Can we add some tests where dtype is not DataType? I'm wondering if nested numpy arrays have any problems.

Reply:
Sure, I guess it works fine but I'll make it sure.

Reply:
Note from offline: With current behavior,`np.array` passed is returned as python list. This is not a bug but ideally we don't make any side effect to the given data. Will work on this in separate follow-up PR.